### PR TITLE
BLS support

### PR DESCRIPTION
**Pull request by rllola**
Rewrite of PR https://github.com/Zondax/filecoin-rs/pull/121

### CODE REVIEW

File: signer/src/lib.rs [hunk 1]
```diff
@@ -26,10 +33,13 @@ pub struct Mnemonic(pub String);
 /// CBOR message in a buffer
 pub struct CborBuffer(pub Vec<u8>);
 
-pub const SIGNATURE_RECOVERY_SIZE: usize = SIGNATURE_SIZE + 1;
+impl AsRef<[u8]> for CborBuffer {
```
Comment (jleni) [inline signer/src/lib.rs:36]: nice :)

File: signer/src/lib.rs [hunk 2]
```diff
@@ -267,20 +288,7 @@ pub fn transaction_sign(
...
-    let cid_hashed = utils::get_digest(&message_cbor.0);
-
-    let message_digest = Message::parse_slice(&cid_hashed)?;
-
-    let (signature_rs, recovery_id) = sign(&message_digest, &secret_key);
-
-    let mut signature = Signature { 0: [0; 65] };
-    signature.0[..64].copy_from_slice(&signature_rs.serialize()[..]);
-    signature.0[64] = recovery_id.serialize();
+    let signature = transaction_sign_raw(unsigned_message, private_key)?;
```
Comment (jleni) [inline signer/src/lib.rs:291]: very nice

File: signer/src/lib.rs [hunk 3]
```diff
@@ -334,18 +335,112 @@ pub fn verify_signature(
...
+    use crate::utils;
+    use bls_signatures;
+    use bls_signatures::Serialize;
+    use forest_address::Address;
+    use rand::{Rng, SeedableRng};
+    use rand_xorshift::XorShiftRng;
+    use rayon::prelude::*;
+
+    const BLS_PUBKEY: &str = "ade28c91045e89a0dcdb49d5ed0d62a4f02d78a96dbd406a4f9d37a1cd2fb5c29058def79b01b4d1556ade74ffc07904";
+    // FIXME! Might be invalid
```
Comment (jleni) [inline signer/src/lib.rs:441]: is this something serious to correct or verify?

File: docs/native/api.md [hunk 4]
```diff
@@ -132,11 +132,17 @@ match transaction {
...
+e.g :
+
+* "t**1**b4zd6ryj5dsnwda5jtjxj6ptkia5e35s52ox7ka" is a [protocol 1 address](https://filecoin-project.github.io/specs/#protocol-1-libsecpk1-elliptic-curve-public-keys) therefore `transaction_sign_raw` will attempt to the sign the transaction using `Secp256k1`.
+
+* "t**3**vxrizeiel2e2bxg3jhk62dlcutyc26fjnw6ua2sptu32dtjpwxbjawg666nqdngrkvvn45h7yb4qiya6ls7q" is a [protocol 3 address](https://filecoin-project.github.io/specs/#protocol-3-bls). Here we will use `BLS` signing scheme.
 
 Arguments:
 * **transaction**: a filecoin transaction;
-* **privatekey**: a private key as hexstring;
+* **privatekey**: a private key as hexstring (should be the associated private key of `from` address of the transaction);
```
Comment (jleni) [inline docs/native/api.md:145]: It is a bit strange to require hexstring in the native rust API but we can later improve this in another issue.

File: signer/src/lib.rs [hunk 5]
```diff
@@ -227,35 +210,73 @@ pub fn transaction_parse(
...
 
+fn transaction_sign_bls_raw(
+    unsigned_message_api: &UnsignedMessageAPI,
+    private_key: &PrivateKey,
+) -> Result<SignatureBLS, SignerError> {
+    let message_cbor = transaction_serialize(unsigned_message_api)?;
+
+    let sk = bls_signatures::PrivateKey::from_bytes(&private_key.0)?;
+
+    // REVIEW: no pre-hashing ?
```
Comment (jleni) [inline signer/src/lib.rs:242]: If this is not clear, I think we should ask the specs or cryptography teams

File: signer/src/lib.rs [hunk 6]
```diff
@@ -334,18 +335,112 @@ pub fn verify_signature(
     Ok(verify(&blob_to_sign, &signature_rs, &public_key))
 }
 
+fn verify_bls_signature(
+    signature: &SignatureBLS,
+    cbor_buffer: &CborBuffer,
+) -> Result<bool, SignerError> {
+    // TODO: need a function to extract from public key from cbor buffer directly
```
Comment (jleni) [inline signer/src/lib.rs:342]: Let's create an issue for this?

File: signer/src/lib.rs [hunk 7]
```diff
@@ -227,35 +210,73 @@ pub fn transaction_parse(
...
+    private_key: &PrivateKey,
+) -> Result<SignatureBLS, SignerError> {
+    let message_cbor = transaction_serialize(unsigned_message_api)?;
+
+    let sk = bls_signatures::PrivateKey::from_bytes(&private_key.0)?;
+
+    // REVIEW: no pre-hashing ?
+    let sig = sk.sign(&message_cbor.0);
+
+    let mut signature = SignatureBLS::try_from(sig.as_bytes())?;
```
Comment (jleni) [inline signer/src/lib.rs:245]: If this is not clear, I think we should ask the specs or cryptography teams

File: signer/src/lib.rs (hunk 3 above)
Comment (rllola) [reply signer/src/lib.rs:441]: This one was generated using `bls-signatures` so I think we are good. But it wasn't working with the bls javascript lib I tried in the wasm tests.

I don't think it is relevant anymore. We discussed that the javascript lib I was using was most certainly using an other standard and we should instead generate test cases from dig's lib.

File: docs/native/api.md (hunk 4 above)
Comment (rllola) [reply docs/native/api.md:145]: Good catch. It is a mistake in the doc. We are requiring bytes array in the code.

File: signer/src/lib.rs (hunk 6 above)
Comment (rllola) [reply signer/src/lib.rs:342]: Yes. This is not a priority but would avoid the parsing before and converting to pubkey.

**Pull request merged**